### PR TITLE
Rdsaad

### PR DESCRIPTION
**Pull request by akallabeth**
@fifthdegree I´ve taken your pr #8760 and refactored the code a bit.
(also links `cJSON` instead of copying the files into the project)

Things that needed to be done:
* update `CI` to have `libcjson-dev` installed on build nodes

### CODE REVIEW

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:156]: You need space for the null terminator. It's part of the PDUs

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:160]: Reinitialize the arg pointer before calling `vsnprintf`

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:209]: `*obj = prop;`

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:227]: You don't want to do this. Let properties be freed when you free the object

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:255]: As above

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:376]: `cJSON_free()` only frees the object pointed to itself. Better use `cJSON_Delete()`

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:380]: Typo... can you spot it? :)

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:404]: `cJSON_Delete()`

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:462]: We hope for it to *not* be null...

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:501]: Here too you missed something(!)

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:745]: As above

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:775]: And one more

**Review by fifthdegree — CHANGES_REQUESTED** [review]
Works fine after tracking down the couple of minor bugs I pointed out. I'm going to try to resist the irrepressible urge of expressing just how much I am not a fan of over-the-top almost paranoid-like checking and rechecking of values you pretty much already know is correct. I am however quite disappointed that you just *had* to typedef an enum. All teasing aside, looks great! I'll close #8760

**Comment by akallabeth** [comment]
> Works fine after tracking down the couple of minor bugs I pointed out. I'm going to try to resist the irrepressible urge of expressing just how much I am not a fan of over-the-top almost paranoid-like checking and rechecking of values you pretty much already know is correct. I am however quite disappointed that you just _had_ to typedef an enum. All teasing aside, looks great! I'll close #8760

yes, I know the feeling. style is something you can only agree or hate :grin: 
thanks for testing (and fixing my oversights)

File: libfreerdp/core/aad.c
Comment (akallabeth) [reply libfreerdp/core/aad.c:462]: why? just fail if something works is much more fun :)
damn, that is a nasty one

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/aad.c
Comment (fifthdegree) [inline libfreerdp/core/aad.c:340]: There are either too many format conversions or too few arguments (same for the one just below)

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test again please...

**Comment by fifthdegree** [comment]
Works correctly now. Would be great if we can find a way to have the code be redirected to us though

**Comment by akallabeth** [comment]
@fifthdegree what do you mean with redirected?

**Comment by fifthdegree** [comment]
@akallabeth No, not at all. I was talking about the oauth2 authorization code. I can see now how that wasn't clear :)

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@mfleisz @hardening @bmiklautz the pull request is ready so far (as in compiles with our ci now and @fifthdegree did some fine function testing)
only remaining question now is should we add this as a required build option or make it optional? (we additionally depend on `cJSON` now)

**Comment by mfleisz** [comment]
Since this is only required for the AAD feature I'd make it an option (but have it enabled on the CI).

**Comment by akallabeth** [comment]
> @akallabeth No, not at all. I was talking about the oauth2 authorization code. I can see now how that wasn't clear :)

ok, then lets get this pr upstream and then lets fix the remaining `how to automatically get the authentication code` part

**Comment by awakecoding** [comment]
@fifthdegree I'm excited about these changes, I've tried the new enablerdsaadauth:i:1 option in mstsc between HAADJ work devices, and it took everything just for the login prompt to finish without crashing, after which I have a permanent connection error and no way to diagnose the issue. I'm hoping FreeRDP would work better or at least give me a better way to see where it fails at the protocol level :)

**Comment by akallabeth** [comment]
@awakecoding all I´m waiting for is a positive review of someone with write access ;) I think you still have them ^^

**Comment by akallabeth** [comment]
and yes, I´m also really excited about all these new things :)

File: libfreerdp/core/nego.c
Comment (awakecoding) [inline libfreerdp/core/nego.c:61]: why did this get bumped to 32? just curious as 16 looks safe enough, but I don't mind extra caution

File: libfreerdp/CMakeLists.txt
Comment (awakecoding) [inline libfreerdp/CMakeLists.txt:115]: thanks - I was wondering if cJSON was going to be a soft dependency, and it is :)

**Review by awakecoding — APPROVED** [review]

File: libfreerdp/core/nego.c
Comment (akallabeth) [reply libfreerdp/core/nego.c:61]: `#define PROTOCOL_RDSAAD 0x00000010` might explain it ;)

File: libfreerdp/CMakeLists.txt
Comment (akallabeth) [reply libfreerdp/CMakeLists.txt:115]: soft as we just terminate the session with a warning if `AAD` is used without it. (settings and command line do not check this)

**Comment by akallabeth** [comment]
@awakecoding and merged, thank you for the revies :)

**Comment by fifthdegree** [comment]
@awakecoding Happy to help! Please do let us know how it works out for you. Also, if you have any ideas on how to automatically get the authorization code, I'm all ears

**Comment by awakecoding** [comment]
> @awakecoding Happy to help! Please do let us know how it works out for you. Also, if you have any ideas on how to automatically get the authorization code, I'm all ears

How are you currently testing? I was thinking of giving it a try on Windows, as I suspect there's a way to just grab the authorization code from the system cache. I've had a developer work on reversing the undocumented parts of the previous protocol (RDP NLA + Azure AD + PKU2U) and implement a working solution on Windows with the help of manually extracted certificates, I haven't had a chance to try and polish it for something functional, but the dev wrote a tool to fetch the required PKU2U certificates from Azure AD, I wouldn't be surprised if this tool needs to get a similar authorization code required for the newer RDS AAD protocol, maybe take a closer look at it and see? https://github.com/Devolutions/sspi-rs/tree/master/tools/aadjoin

**Pull request merged**
